File: zigzagConversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows <= 1 or len(s) <= 2:
            return s
        
        res = ""
        row = 0
        def skippedCal(r):
            if r == 0:
                return r
            if r == 1:
                return 2
            return (r - 1) * 2 + 2
        for r in range(numRows):
            c = r
            skip = skippedCal(numRows - r - 1)
            reverse = skippedCal(r)
            forward = True
            while c < len(s):
                print(s[c], forward, skip, reverse)
                res += s[c]
                if not skip:
                    c += reverse
                    forward = False
                elif not reverse:
                    c += skip
                    forward = True
                elif forward:
                    c += skip
                    forward = False
                elif not forward:
                    c += reverse
                    forward = True      
        return res
                

File: zigzagconversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if(numRows == 1) return s;
        int L = s.length();
        string res = "";
        int MAX = 2 * (numRows - 1);
        for (int r = 0; r < numRows; r++) {
            int start = r;
            int hop = MAX - 2 * r;
            int revhop = MAX - hop;
            if(hop == 0) {
                hop = MAX;
            } else if(revhop == 0) {
                revhop = MAX;
            }
            while (start < L) {
                res += s[start];
                start += hop;
                if (start >= L) break;
                res += s[start];
                start += revhop;
            }
        }
        return res;
    }
};
